clang-tools-extra/clang-doc/JSONGenerator.cpp

```diff
@@ -0,0 +1,352 @@
+  Obj["IsUsing"] = I.IsUsing;
+  Object TypeObj = Object();
+  serializeInfo(I.Underlying, TypeObj);
+  Obj["Underlying"] = std::move(TypeObj);
+}
+
+static void serializeInfo(const RecordInfo &I, json::Object &Obj,
+                          std::optional<StringRef> RepositoryUrl) {
+  serializeCommonAttributes(I, Obj, RepositoryUrl);
+  Obj["FullName"] = I.Name.str();
+  Obj["TagType"] = getTagType(I.TagType);
+  Obj["IsTypedef"] = I.IsTypeDef;
+
+  if (!I.Children.Functions.empty()) {
+    json::Value PublicFunctionArr = Array();
+    json::Array &PublicFunctionARef = *PublicFunctionArr.getAsArray();
+    json::Value ProtectedFunctionArr = Array();
+    json::Array &ProtectedFunctionARef = *ProtectedFunctionArr.getAsArray();
+
+    for (const auto &Function : I.Children.Functions) {
+      json::Object FunctionObj;
+      serializeInfo(Function, FunctionObj, RepositoryUrl);
+      AccessSpecifier Access = Function.Access;
+      if (Access == AccessSpecifier::AS_public)
+        PublicFunctionARef.push_back(std::move(FunctionObj));
+      else if (Access == AccessSpecifier::AS_protected)
+        ProtectedFunctionARef.push_back(std::move(FunctionObj));
+    }
+
+    if (!PublicFunctionARef.empty())
+      Obj["PublicFunctions"] = std::move(PublicFunctionArr);
+    if (!ProtectedFunctionARef.empty())
+      Obj["ProtectedFunctions"] = std::move(ProtectedFunctionArr);
+  }
+
+  if (!I.Members.empty()) {
+    json::Value PublicMembers = Array();
+    json::Array &PubMemberRef = *PublicMembers.getAsArray();
+    json::Value ProtectedMembers = Array();
+    json::Array &ProtMemberRef = *ProtectedMembers.getAsArray();
+
+    for (const MemberTypeInfo &Member : I.Members) {
+      json::Object MemberObj = Object();
+      MemberObj["Name"] = Member.Name;
+      MemberObj["Type"] = Member.Type.Name;
+
+      if (Member.Access == AccessSpecifier::AS_public)
+        PubMemberRef.push_back(std::move(MemberObj));
+      else if (Member.Access == AccessSpecifier::AS_protected)
+        ProtMemberRef.push_back(std::move(MemberObj));
+    }
+
+    if (!PubMemberRef.empty())
+      Obj["PublicMembers"] = std::move(PublicMembers);
+    if (!ProtMemberRef.empty())
+      Obj["ProtectedMembers"] = std::move(ProtectedMembers);
+  }
+
+  serializeCommonChildren(I.Children, Obj, RepositoryUrl);
+}
+
+Error JSONGenerator::generateDocs(
+    StringRef RootDir, llvm::StringMap<std::unique_ptr<doc::Info>> Infos,
+    const ClangDocContext &CDCtx) {
+  StringSet<> CreatedDirs;
+  StringMap<std::vector<doc::Info *>> FileToInfos;
+  for (const auto &Group : Infos) {
+    Info *Info = Group.getValue().get();
+
+    SmallString<128> Path;
+    sys::path::native(RootDir, Path);
+    sys::path::append(Path, Info->getRelativeFilePath(""));
+    if (!CreatedDirs.contains(Path)) {
+      if (std::error_code Err = sys::fs::create_directories(Path);
+          Err != std::error_code())
+        ExitOnErr(createFileError(Twine(Path), Err));
```

Comment:
Should this just return the error and let the top level handler in the tool deal with it?